source/bigint.c

```diff
@@ -244,82 +303,69 @@ struct aws_bigint *aws_bigint_new_from_hex(struct aws_allocator *allocator, stru
     return NULL;
 }
 
-struct aws_bigint *aws_bigint_new_from_int64(struct aws_allocator *allocator, int64_t value) {
-    AWS_PRECONDITION(allocator);
+struct aws_bigint *aws_bigint_new_from_binary_cursor(
+    struct aws_allocator *allocator,
+    struct aws_byte_cursor source,
+    size_t desired_bit_width) {
+    AWS_FATAL_PRECONDITION(allocator);
+    AWS_FATAL_PRECONDITION(aws_byte_cursor_is_valid(&source));
 
-    if (value >= 0) {
-        return aws_bigint_new_from_uint64(allocator, (uint64_t)value);
+    if (source.len == 0) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+        return NULL;
     }
 
-    struct aws_bigint *bigint = NULL;
-    if (value == INT64_MIN) {
-        /* We can't just negate and cast, so just submit a constant */
-        bigint = aws_bigint_new_from_uint64(allocator, INT64_MIN_AS_HEX);
-    } else {
-        /* The value is negative but can be safely negated to a positive value before casting to uint64 */
-        bigint = aws_bigint_new_from_uint64(allocator, (uint64_t)(-value));
+    size_t digit_count = (source.len - 1) / BYTES_PER_BASE_DIGIT + 1;
+    if (desired_bit_width == 0) {
+        desired_bit_width = digit_count * BASE_BITS;
     }
 
-    if (bigint == NULL) {
+    size_t desired_digit_count = s_convert_bit_width_to_digit_count(desired_bit_width);
+    if (digit_count > desired_digit_count) {
+        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
         return NULL;
     }
 
-    bigint->sign = -1;
-
-    AWS_POSTCONDITION(aws_bigint_is_valid(bigint));
-
-    return bigint;
-}
-
-struct aws_bigint *aws_bigint_new_from_uint64(struct aws_allocator *allocator, uint64_t value) {
-    AWS_PRECONDITION(allocator);
-
-    struct aws_bigint *bigint = aws_mem_calloc(allocator, 1, sizeof(struct aws_bigint));
+    struct aws_bigint *bigint = aws_bigint_new_zero(allocator, desired_bit_width);
     if (bigint == NULL) {
         return NULL;
     }
 
-    bigint->allocator = allocator;
+    for (size_t i = 0; i < source.len; ++i) {
```

Comment:
Why not just loop from source.len - 1 to 0?

Reply:
Natural mistrust of iterating downwards with an unsigned counter + would have to invert the i references anyways.